Add config flow to nederlandse_spoorwegen

heindrichpaul opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This PR enhances the Nederlandse Spoorwegen integration by implementing a full UI configuration flow and raising the integration quality to meet the Silver tier requirements. It adds robust reauthentication and reconfiguration support, improves error handling, and ensures a better user experience through clear validation and messaging. These changes make the integration easier to set up, maintain, and use within Home Assistant.

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes mypy errors and improves config flow robustness
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/39826
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to requirements_all.txt.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Similar code found with 1 license type

heindrichpaul committed:
Adding tests to the existing NS integration

heindrichpaul committed:
Using a fixed now in the tests for stability

heindrichpaul committed:
Adding tests for the existing sensors

heindrichpaul committed:
Added final tests

heindrichpaul committed:
Added config flow

heindrichpaul committed:
Added config flow

heindrichpaul committed:
Merge branch 'feature/improving_NS_integration' of github.com:heindrichpaul/core into feature/improving_NS_integration

heindrichpaul committed:
Added tests and UI config flow to integration

heindrichpaul committed:
Fixed config flow

heindrichpaul committed:
Fixed test

heindrichpaul committed:
Fixed translation. Added more tests

heindrichpaul committed:
Merge pull request #2 from heindrichpaul/feature/improving_NS_integration

Feature/improving ns integration

heindrichpaul committed:
Merge branch 'home-assistant:dev' into dev

heindrichpaul committed:
Added unique entity id's

heindrichpaul committed:
updated the quality status

heindrichpaul committed:
Updated the quality status

heindrichpaul committed:
Updated the quality status

heindrichpaul committed:
Merge branch 'home-assistant:dev' into dev

heindrichpaul committed:
Merge pull request #3 from heindrichpaul/feature/improving_NS_integration

Feature/improving ns integration

heindrichpaul committed:
Merge pull request #4 from heindrichpaul/dev

a

heindrichpaul committed:
Updated the config_flow to use const.py constants

MartinHjelmare commented on homeassistant/components/nederlandse_spoorwegen/quality_scale.yaml:
```diff
@@ -0,0 +1,133 @@
+# Home Assistant Integration Code Quality Rules for Nederlandse Spoorwegen
```
Please update the quality scale in a separate PR.

heindrichpaul replied:
I have removed it from this PR will add it to another

heindrichpaul committed:
Refactor Nederlandse Spoorwegen integration: update manifest formatting and remove quality scale file, Will create another PR to add this

heindrichpaul committed:
Fixed order

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+        "title": "Nederlandse Spoorwegen Setup",
+        "description": "Set up your Nederlandse Spoorwegen integration.",
+        "data": {
+          "api_key": "API key"
```
Can use a common string here:
```suggestion
          "api_key": "[%key:common::config_flow::data::api_key%]"
```

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+        }
+      },
+      "routes": {
+        "title": "Add Route",
```
Sentence-casing for all strings:

```suggestion
        "title": "Add route",
```

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Nederlandse Spoorwegen Setup",
```
Sentence-casing here:

```suggestion
        "title": "Nederlandse Spoorwegen setup",
```

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+        "title": "Re-authenticate",
+        "description": "Your NS API key needs to be updated.",
+        "data": {
+          "api_key": "API key"
```
```suggestion
          "api_key": "[%key:common::config_flow::data::api_key%]"
```

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+        "title": "Reconfigure",
+        "description": "Update your NS API key.",
+        "data": {
+          "api_key": "API key"
```
```suggestion
          "api_key": "[%key:common::config_flow::data::api_key%]"
```

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+    "error": {
+      "cannot_connect": "Could not connect to NS API. Check your API key.",
+      "invalid_auth": "Invalid API key.",
+      "unknown": "Unknown error occurred."
```
```suggestion
      "unknown": "[%key:common::config_flow::error::unknown%]"
```

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+  "options": {
+    "step": {
+      "init": {
+        "title": "Configure Routes",
```
```suggestion
        "title": "Configure routes",
```

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+    },
+    "error": {
+      "cannot_connect": "Could not connect to NS API. Check your API key.",
+      "invalid_auth": "Invalid API key.",
+      "unknown": "Unknown error occurred.",
```
```suggestion
      "invalid_auth": "[%key:common::config_flow::error::invalid_api_key%]",
      "unknown": "[%key:common::config_flow::error::unknown%]",
```

NoRi2909 reviewed (commented):
A quick first look over the strings from a translator's perspective.

Note that the `data` and `data_description` blocks show up three times in the strings.
Therefore the 2nd and 3rd block should be replaced with references from the 1st.

heindrichpaul committed:
Refactor Nederlandse Spoorwegen integration: standardize titles and descriptions in strings.json

heindrichpaul committed:
Merge branch 'home-assistant:dev' into dev

heindrichpaul committed:
Merge branch 'dev' of github.com:heindrichpaul/core into feature/improving_NS_integration

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+        "title": "Add route",
+        "description": "[%key:component::nederlandse_spoorwegen::config::step::routes::title%]",
+        "data": {
+          "name": "Route name",
+          "from": "From station",
+          "to": "To station",
+          "via": "Via station (optional)",
+          "time": "Departure time (optional)"
+        },
+        "data_description": {
+          "name": "A name for this route.",
+          "from": "Departure station code (e.g., AMS)",
+          "to": "Arrival station code (e.g., UTR)",
+          "via": "Optional via station code (e.g., RTD)",
+          "time": "Optional departure time in HH:MM:SS (24h)"
```
Let's deduplicate all those, too:

```suggestion
          "name": "[%key:component::nederlandse_spoorwegen::config::step::routes::data::name%]",
          "from": "[%key:component::nederlandse_spoorwegen::config::step::routes::data::from%]",
          "to": "[%key:component::nederlandse_spoorwegen::config::step::routes::data::to%]",
          "via": "[%key:component::nederlandse_spoorwegen::config::step::routes::data::via%]",
          "time": "[%key:component::nederlandse_spoorwegen::config::step::routes::data::time%]"
        },
        "data_description": {
          "name": "[%key:component::nederlandse_spoorwegen::config::step::routes::data_description::name%]",
          "from": "[%key:component::nederlandse_spoorwegen::config::step::routes::data_description::from%]",
          "to": "[%key:component::nederlandse_spoorwegen::config::step::routes::data_description::to%]",
          "via": "[%key:component::nederlandse_spoorwegen::config::step::routes::data_description::via%]",
          "time": "[%key:component::nederlandse_spoorwegen::config::step::routes::data_description::time%]"
```

Same for that identical block below.

heindrichpaul committed:
Update homeassistant/components/nederlandse_spoorwegen/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Update homeassistant/components/nederlandse_spoorwegen/strings.json

Removed duplications

Co-authored-by: Norbert Rittel <norbert@rittel.de>

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

NoRi2909 commented on homeassistant/components/nederlandse_spoorwegen/strings.json:
```diff
@@ -0,0 +1,144 @@
+        }
+      },
+      "add_route": {
+        "title": "Add route",
+        "description": "[%key:component::nederlandse_spoorwegen::config::step::routes::title%]",
```
Looks like `title` and `description` got swapped here with the single reference, and we can reference both, too:

```suggestion
        "title": "[%key:component::nederlandse_spoorwegen::config::step::routes::title%]",
        "description": "[%key:component::nederlandse_spoorwegen::config::step::routes::description%]",
```

heindrichpaul committed:
Update homeassistant/components/nederlandse_spoorwegen/strings.json

Co-authored-by: Norbert Rittel <norbert@rittel.de>

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
removing un needed logging

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

heindrichpaul committed:
removing debugging logs

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

heindrichpaul committed:
Refactor Nederlandse Spoorwegen integration setup and improve error handling

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul commented:
@joostlek, I had contact with @Martreides and I told him I would look into his change to use sub-entities but will include it in a new PR as I do not want this one to become too big. As it is already on the larger side.

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Add unit tests for Nederlandse Spoorwegen integration

- Implement tests for NSTripSensor to verify state when all trips are in the past.
- Create comprehensive service tests for add_route and remove_route functionalities.
- Include scenarios for service calls when no integration is configured or when the integration is not loaded.
- Mock necessary components and API responses to ensure isolated testing.

heindrichpaul committed:
Cleaned up code

heindrichpaul committed:
Remove unused stations attribute from coordinator and related tests

heindrichpaul committed:
Cleaned strings.json

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul commented:
@joostlek I added the sub-entities and further improved the config flow.

heindrichpaul committed:
Reverted

heindrichpaul committed:
Merge branch 'feature/improving_NS_integration' of github.com:heindrichpaul/core into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

joostlek commented on homeassistant/components/nederlandse_spoorwegen/sensor.py:
```diff
@@ -2,289 +2,215 @@
+    if coordinator is None:
+        _LOGGER.error("Coordinator not found in runtime_data for NS integration")
         return
-
-    sensors = []
-    for departure in config.get(CONF_ROUTES, {}):
-        if not valid_stations(
-            stations,
-            [departure.get(CONF_FROM), departure.get(CONF_VIA), departure.get(CONF_TO)],
-        ):
-            continue
-        sensors.append(
-            NSDepartureSensor(
-                nsapi,
-                departure.get(CONF_NAME),
-                departure.get(CONF_FROM),
-                departure.get(CONF_TO),
-                departure.get(CONF_VIA),
-                departure.get(CONF_TIME),
+    _LOGGER.debug(
+        "NS sensor setup: coordinator=%s, entry_id=%s", coordinator, entry.entry_id
+    )
+
+    # Always create the service sensor
+    entities: list[SensorEntity] = [NSServiceSensor(coordinator, entry)]
+
+    # Create trip sensors for each route
+    if coordinator.data and "routes" in coordinator.data:
+        for route_key, route_data in coordinator.data["routes"].items():
+            route = route_data["route"]
+            entities.append(
+                NSTripSensor(
+                    coordinator,
+                    entry,
+                    route,
+                    route_key,
+                )
             )
-        )
-    add_entities(sensors, True)
```
We should import the old configuration so the user only has to remove their old config and they'd be ready.

joostlek commented:
Oh and please don't merge `dev` in your branch that often, currently the CI doesn't run because you haven't merged anything yet, but after that it would cost quite a bit of CI resources :)

We generally only do that when it has been out of sync for a LONG time, or when we know it will break, or for other reasons

heindrichpaul commented on homeassistant/components/nederlandse_spoorwegen/config_flow.py:
```diff
@@ -0,0 +1,491 @@
+                    options={"routes": []},
+                )
+        data_schema = vol.Schema({vol.Required(CONF_API_KEY): str})
+        return self.async_show_form(
+            step_id="user",
+            data_schema=data_schema,
+            errors=errors,
+            description_placeholders={"station_list_url": STATION_LIST_URL},
+        )
```
It is actually already part of this change. The link is still there if the drop down doesn't load it is a text field and then you will need to enter the code.

joostlek replied:
Yes but we need to prevent the drop down from not loading and always give the user a nice way instead of providing a backup :)

heindrichpaul commented on homeassistant/components/nederlandse_spoorwegen/config_flow.py:
```diff
@@ -0,0 +1,491 @@
+                else:
+                    errors["base"] = "cannot_connect"
+            if not errors:
+                await self.async_set_unique_id(api_key)
```
Yes, NS API keys can change - users can regenerate them in the NS API portal and they typically have multiple keys they can rotate between. I have added a more stable unique ID so that HA doesn't see it as a new integration when it does change.

heindrichpaul committed:
Refactor Nederlandse Spoorwegen integration tests to utilize NSRuntimeData for runtime data management, enhancing structure and readability. Added comprehensive subentry flow tests for route management, including validation, error handling, and successful route creation scenarios. Improved test coverage for options flow and reconfiguration processes, ensuring robust handling of user inputs and station data.

heindrichpaul committed:
Merge branch 'feature/improving_NS_integration' of github.com:heindrichpaul/core into feature/improving_NS_integration

heindrichpaul committed:
Add migration tests for legacy routes to subentries in Nederlandse Spoorwegen integration

heindrichpaul committed:
Add investigation scripts and tests for Nederlandse Spoorwegen integration

- Introduced , , and  to analyze device and entity registries post-migration.
- Implemented JSON error handling tests in  to ensure graceful handling of JSON parsing errors.
- Updated  to remove obsolete route service tests and ensure proper functionality of the coordinator.
- Refactored  to simplify async setup tests and removed service registration assertions.
- Enhanced  with a new test for device association after migration, ensuring correct entity and device creation.
- Deleted  as its functionality is now covered in other tests.
- Improved  to validate station options formatting and ensure correct handling of various station formats.

heindrichpaul committed:
Refactor Nederlandse Spoorwegen tests and migration logic

- Updated test_migration.py to use AsyncMock for NSAPIWrapper and refactored legacy route migration tests.
- Added new test_migration_entry.py to cover config entry migration scenarios.
- Removed outdated tests from test_sensor.py related to service and trip sensors, adapting to new architecture.
- Introduced test_station_parsing.py to validate station string parsing functionality.
- Deleted test_sensor_past_trips.py as it was no longer relevant.
- Updated test_subentry_flow.py to ensure time format consistency in route configurations.

heindrichpaul committed:
Refactor Nederlandse Spoorwegen config flow by removing reauthentication and reconfiguration steps. Update strings for improved clarity and consistency in route management.

heindrichpaul committed:
Add comprehensive tests for Nederlandse Spoorwegen sensors and utilities

- Implement tests for sensor logic in test_sensor.py, covering various scenarios including device association, availability, and native value handling.
- Create new test files: test_sensor_edge_cases.py and test_sensor_new.py for additional edge case testing and new sensor functionalities.
- Introduce utility tests in test_utils.py to validate time formatting, route validation, and safe data extraction methods.
- Ensure coverage for scenarios with missing or invalid data in both sensors and utilities.

heindrichpaul committed:
Refactor Nederlandse Spoorwegen integration files for improved structure and maintainability

heindrichpaul committed:
Refactor Nederlandse Spoorwegen diagnostics and test files for improved clarity and maintainability

heindrichpaul committed:
Refactor Nederlandse Spoorwegen setup logic for improved error handling and code clarity

heindrichpaul committed:
Add comment to clarify parallel update limit for Nederlandse Spoorwegen API

joostlek commented on homeassistant/components/nederlandse_spoorwegen/__init__.py:
```diff
@@ -1 +1,197 @@
+    """Handle removal of a config entry."""
+
+
+async def _async_migrate_legacy_routes(
+    hass: HomeAssistant, entry: NSConfigEntry
+) -> None:
+    """Migrate legacy routes from configuration data into subentries.
+
+    This handles routes stored in entry.data[CONF_ROUTES] from legacy YAML config.
+    One-time migration to avoid duplicate imports.
+    """
```
Yes and no, you're running that code in `async_setup_entry`, which means that there already should be an entry to reach that code.

```py
def setup_platform(
    hass: HomeAssistant,
    config: ConfigType,
    add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType | None = None,
) -> None:
```
This is the current entrypoint of the integration and we should use this (or the async counterpart) to start an import flow, and then the code in `async_step_import` in the config flow should make sure all the right subentries are created.

heindrichpaul replied:
@joostlek I have now added the logic to the async_setup_entry as well. I have also run multiple tests locally and it imports the routes and config without issues

heindrichpaul committed:
adding the import flow to the async setup

heindrichpaul committed:
Fixed the import to use the platform_setup

heindrichpaul commented:
> You should use `async_setup_platform` in `sensor.py` instead of `async_setup` as we should keep the current entrypoint for YAML

It is now fixed. Sorry had the wrong config missed the sensor and platform that was why I didn't see it.

joostlek commented:
Will check!

heindrichpaul commented:
@joostlek did you have time to take a look? I wanted to update the docs PR but am first waiting on this one's go ahead before I finish that.

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration

heindrichpaul committed:
Merge branch 'dev' into feature/improving_NS_integration
